docs/development/room-dag-concepts.md

```diff
@@ -0,0 +1,83 @@
+# Room DAG concepts
+
+## Edges
+
+The word "edge" comes from graph theory lingo. An edge is just a connection
+between two events. In Synapse, we connect events by specifying their
+`prev_events`. A subsequent event points back at a previous event.
+
+```
+A (oldest) <---- B <---- C (most recent)
+```
+
+
+## Depth and stream ordering
+
+Events are sorted by `(topological_ordering, stream_ordering)` where
+`topological_ordering` is just `depth`. In other words, we first sort by `depth`
+and then tie-break based on `stream_ordering`. `depth` is incremented as new
+messages are added to the DAG. Normally, `stream_ordering` is an auto
+incrementing integer but for `backfilled=true` events, it decrements.
+
+`depth` is not re-calculated if a message is inserted into the middle of the DAG.
```

Comment:
`depth` is an inherent property of the event - it's calculated by the originating server as `max(e.depth for e in prev_events)+1`, and it is then covered by the signature (and therefore set in stone).

Reply:
Woah! I was under the impression that it was a just a field that Synapse calculated to make the slicing and dicing of all events easier.

From the client perspective it doesn't even exist: https://spec.matrix.org/unstable/client-server-api/ 

But I do see some mentions from the server API: https://spec.matrix.org/unstable/server-server-api/ 

Why don't I see when I inspect the original source of an event in Element for example? Where is it stored on the event? Any docs?

---

The reason I mentioned that line in the first place is because during the MSC2716 talks, there was some discussion of doing some major refactoring so that the `depth` is calculated. Maybe this was meant for a future MSC vs a homeserver implementation 😕

> So having rediscussed this IRL: Dendrite (and Synapse) currently get their `depth` parameters used for ordering from the wire. Ideally, we'd calculate the depth parameter instead
>
> https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r559828899

Relates to https://github.com/matrix-org/matrix-doc/issues/1229

Reply:
> Why don't I see when I inspect the original source of an event in Element for example? 

Well, therein lies a complication. The format of events served up to clients is not quite the same as those that get passed around between homeservers. I opened https://github.com/matrix-org/matrix-doc/issues/3305 to clarify this.



> The reason I mentioned that line in the first place is because during the MSC2716 talks, there was some discussion of doing some major refactoring so that the `depth` is calculated.

Well yes, it would be nice if we didn't have to trust remote homeservers to tell us the depth of an event, either by calculating it ourselves (somehow) or, more likely, avoiding our dependence on it. But that's all a bit sci-fi at the moment, and if you're writing Synapse developer docs, I think it's better to make them reflect current reality than future aspirations.